util/dvsim/LsfLauncher.py

```diff
@@ -0,0 +1,379 @@
+    # read it so we retry on the next poll, no more than 10 times.
+    max_poll_retries = 10
+
+    # TODO: Add support for build/run/cov job specific resource requirements:
+    #       cpu, mem, disk, stack.
+    # TODO: Allow site-specific job resource usage setting using
+    #       `DVSIM_LSF_CFG` environment variable.
+
+    def __init__(self, deploy):
+        super().__init__(deploy)
+
+        # Set the status. Only update after the job is done - i.e. status will
+        # transition from None to P/F/K.
+        self.status = None
+
+        # Maintain the job script output as an instance variables for polling
+        # and cleanup.
+        self.job_script_out = None
+
+        # If we already opened the job script output file (but have not
+        # determined the outcome), then we maintain the file descriptor rather
+        # then reopening it and starting all over again on the next poll.
+        self.job_script_out_fd = None
+        self.job_script_out_err_msg = []
+        self.job_script_out_err_msg_found = False
+
+        # Set the job id.
+        self.job_id = None
+
+        # Polling retry counter..
+        self.num_poll_retries = 0
+
+    @staticmethod
+    def prepare_workspace_for_cfg(cfg):
+        '''This is a static method to prepare the workspace before launching
+        any job using this method, but specific to a config.
+        '''
+
+        # Create the job dir.
+        LsfLauncher.jobs_dir[cfg] = Path(cfg.scratch_path, "lsf",
+                                         cfg.timestamp)
+        clean_odirs(odir=LsfLauncher.jobs_dir[cfg], max_odirs=2)
+        os.makedirs(Path(LsfLauncher.jobs_dir[cfg]), exist_ok=True)
+
+        LsfLauncher.jobs[cfg] = {}
+
+    @staticmethod
+    def make_job_array_script_text(cfg, job_name):
+        """Creates the job array script text.
+
+        Once all job in the array are dispatched, the job array script is
+        constructed. It is a bash script that takes the job index as a single
+        argument. This index is set in the bsub command as '$LSB_JOBINDEX',
+        which bsub sets as the actual index when launching that job in the
+        array. This script is super simple - it is just a giant case statement
+        that switches on the job index to run that specific job.
+
+        Returns an iterable representing the lines of the script.
+        """
+        lines = ["#!/usr/bin/env bash\n", "case $1 in\n"]
+
+        for job in LsfLauncher.jobs[cfg][job_name]:
+            # Redirect the job's stdout and stderr to it's log file.
+            cmd = "{} > {} 2>&1".format(job.deploy.cmd,
+                                        job.deploy.get_log_path())
+
+            # Activate and de-activate the python virtualenv if it exists.
+            if Launcher.python_venv:
+                cmd = "source {}/bin/activate && {} && deactivate".format(
+                    Launcher.python_venv, cmd)
```

Comment:
Oh! I read this as "activate; deactivate" rather than "activate; do_something; deactivate". Now I understand, thanks!

Thinking about it, I wonder whether the whole script could run within the venv? In that case, you could put the call to activate at the start of the script and you don't need the deactivate at all, leading to a slightly smaller script. No biggie, though.